fix: ignore cycles of length 1 in dependency graph when sorting

mdrohmann opened the pull request:
Cycles of length 1 in the dependency graph ie., packages that depend on
themselves, confuse the 'graph-sequencer' function that is used to order the
packages topologically.

Sometimes, it makes sense to make packages depend on themselves eg., to install
their exposed scripts in `node_modules/.bin`, so we should either fix the
`graph-sequencer` or apply this patch.

An example for this behavior is:

- Package 'a' depends on 'b' and 'c'
- Package 'b' depends on 'b'
- Package 'c' depends on 'b' and 'c'

Now run:

	pnpm recursive --filter=a... [some command]

**Expected behavior**:

- Packages are visited in order [['b'], ['c'], ['a']], where no parallel
  execution is possible.

**Actual behavior**:

- Packages are visited in order [['b'], ['a'], ['c']].

This patch fixes that behavior.

zkochan commented:
looks ok but a test needs to be added

I think the sequence of building packages is covered here https://github.com/pnpm/pnpm/tree/master/packages/pnpm/test/monorepo or here https://github.com/pnpm/pnpm/tree/master/packages/pnpm/test/recursive

mdrohmann commented:
Ok.

I added a regression test to packages/pnpm/test/monorepo. Does that suffice?  There are other tests that check for the correct topological order already and they still pass with my change.  So, IMO it should be okay.

However, when writing the test, I noticed however, that the 'graph-sequencer' package is also used in `packages/headless/src/runDependenciesScripts.ts`, `packages/supi/src/rebuild/index.ts` and `packages/supi/src/install/index.ts`.  I am not sure if these files need changes too.  In the regression test, I test `pnpm install` and `pnpm recursive test` and in both times, the chunks are computed by sortPackage function that I fixed...

zkochan commented:
perfect! Thanks for your contribution!
